src/hdbscan.jl

```diff
@@ -0,0 +1,303 @@
+        end
+    end
+    return minspantree
+end
+
+function hdbscan_clusters(mst::AbstractVector{MSTEdge}, min_size::Integer)
+    n = length(mst) + 1
+    cost = 0
+    uf = UnionFind(n)
+    clusters = [HdbscanCluster(min_size > 1 ? Int[i] : Int[]) for i in 1:n]
+    sort!(mst)
+    
+    for i in 1 : n-1
+        j, k, c = expand(mst[i])
+        cost += c
+        λ = 1 / cost
+        #child clusters
+        c1 = group(uf, j)
+        c2 = group(uf, k)
+        #reference to the parent cluster
+        clusters[c1].parent = clusters[c2].parent = n+i
+        nc1, nc2 = isnoise(clusters[c1]), isnoise(clusters[c2])
+        if !(nc1 || nc2)
+            #compute stability
+            increment_stability(clusters[c1], λ)
+            increment_stability(clusters[c2], λ)
+            #unite cluster
+            unite!(uf, j, k)
+            #create parent cluster
+            points = members(uf, group(uf, j))
+            push!(clusters, HdbscanCluster(points))
+        elseif !(nc1 && nc2)
+            if nc2 == true
+                (c1, c2) = (c2, c1)
+            end
+            #record the lambda value
+            append!(clusters[c2].λp, fill(λ, length(clusters[c1])))
+            #unite cluster
+            unite!(uf, j, k)
+            #create parent cluster
+            points = members(uf, group(uf, j))
+            push!(clusters, HdbscanCluster(points))
+        else
+            #unite the noise cluster
+            unite!(uf, j, k)
+            #create parent cluster
+            points = members(uf, group(uf, j))
+            if length(points) < min_size
+                push!(clusters, HdbscanCluster(Int[]))
+            else
+                push!(clusters, HdbscanCluster(points))
+            end
+        end
+    end
+    @assert length(clusters) == 2n - 1
+    return clusters
+end
+
+function prune_cluster!(hierarchy::Vector{HdbscanCluster})
+    for i in 1 : length(hierarchy)-1
+        if isnoise(hierarchy[i])
+            c = hierarchy[i]
+            push!(hierarchy[c.parent].children, i)
+            hierarchy[c.parent].children_stability += c.stability
+        else
+            c = hierarchy[i]
+            if c.stability > c.children_stability
+                push!(hierarchy[c.parent].children, i)
+                hierarchy[c.parent].children_stability += c.stability
+            else
+                append!(hierarchy[c.parent].children, c.children)
+                hierarchy[c.parent].children_stability += c.children_stability
+            end
+        end
+    end
+end
+
+# Below are utility functions for building hierarchical trees
+heappush!(h, v) = insert!(h, searchsortedfirst(h, v), v)
+
+# Union-Find
+# structure for managing disjoint sets
+# This structure tracks which sets the elements of a set belong to,
+# and allows us to efficiently determine whether two elements belong to the same set.
+mutable struct UnionFind
+    parent:: Vector{Integer}  # parent[root] is the negative of the size
```

Comment:
`Integer` is an abstract type, so the performance of `Vector{Integer}` is suboptimal, because Julia has to do runtime type inference for each access to its elements.

```suggestion
    parent::Vector{Int}  # parent[root] is the negative of the size
```